### PR TITLE
use jac for the ShampineCollocationInit nlsolve

### PR DESCRIPTION
**Pull request by oscardssmith**
This should improve speed and numerical reliability of the nonlinear solve.

### CODE REVIEW

File: src/initialize_dae.jl [hunk 1]
```diff
@@ -87,12 +87,6 @@ function _initialize_dae!(integrator, prob::DAEProblem,
     end
 end
 
-## NoInit
-
-function _initialize_dae!(integrator, prob::Union{ODEProblem, DAEProblem},
-                          alg::NoInit, x::Union{Val{true}, Val{false}})
```
Comment (ChrisRackauckas) [inline src/initialize_dae.jl:93]: This requires a bump of the minimum.

File: src/initialize_dae.jl [hunk 2]
```diff
@@ -152,20 +146,32 @@ function _initialize_dae!(integrator, prob::ODEProblem, alg::ShampineCollocation
 
         nlequation! = @closure (out, u, p) -> begin
             update_coefficients!(M, u, p, t)
-            #M * (u-u0)/dt - f(u,p,t)
+            # f(u,p,t) + M * (u0 - u)/dt
```
Comment (ChrisRackauckas) [inline src/initialize_dae.jl:149]: why the sign flip?

**Comment by ChrisRackauckas** [comment]
Add some tests?

File: src/initialize_dae.jl (hunk 2 above)
Comment (oscardssmith) [reply src/initialize_dae.jl:149]: the sign flip made the Jacobian a little cleaner IMO. I could revert if you prefer.

File: src/initialize_dae.jl [hunk 3]
```diff
@@ -152,20 +146,32 @@ function _initialize_dae!(integrator, prob::ODEProblem, alg::ShampineCollocation
...
         end
 
-        nlsolve = default_nlsolve(alg.nlsolve, isinplace, u0, isad)
+        jac = if isnothing(f.jac)
+            f.jac
+        else
+            @closure (J, u, p) -> begin
+                # f(u,p,t) + M * (u0 - u)/dt
+                # df(u,p,t)/du - M/dt
+                f.jac(J, u, p, t)
```
Comment (oxinabox) [inline src/initialize_dae.jl:164]: for a DAE, `f.jac` takes a gamma argument.
I suspect this should be:
```suggestion
                f.jac(J, u, p, 0.0, t)
```

**Comment by ChrisRackauckas** [comment]
format and add tests

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
https://buildkite.com/julialang/ordinarydiffeq-dot-jl/builds/2597#0188b521-4359-4210-9420-0c870f644166/835-1252 the test failures look real

**Comment by oscardssmith** [comment]
@ChrisRackauckas good call. Silly typo.

**Comment by ChrisRackauckas** [comment]
Test tolerance may need an update there. Is that within chosen tolerance?

**Comment by oscardssmith** [comment]
The thing that makes no sense to me is that the failing test doesn't have a jacobian, so I'm very confused as to why this is causing a difference. The initialization is failing, but unless passing `jac=nothing` into the nonlinear function is changing something from not passing a `jac`, it really should be doing exactly the same thing as before.

**Comment by oscardssmith** [comment]
Turns out the tests are actually pretty smart. I had a sign error :)

**Pull request merged**